blobfuse/OAuthTokenCredentialManager.cpp

```diff
@@ -0,0 +1,226 @@
+/// </summary>
+std::shared_ptr<OAuthTokenCredentialManager> GetTokenManagerInstance(std::function<OAuthToken(std::shared_ptr<CurlEasyClient>)> defaultCallback) {
+    if(TokenManagerSingleton == nullptr) {
+        if (defaultCallback == nullptr) {
+            throw std::runtime_error("Tried to initialize the OAuthTokenCredentialManager, but failed because the default callback was empty!");
+        }
+
+        TokenManagerSingleton = std::make_shared<OAuthTokenCredentialManager>(defaultCallback);
+    }
+
+    return TokenManagerSingleton;
+}
+
+/// <summary>
+/// OauthTokenCredentialManager Constructor
+/// </summary>
+OAuthTokenCredentialManager::OAuthTokenCredentialManager(
+    std::function<OAuthToken(std::shared_ptr<CurlEasyClient>)> refreshCallback)
+{
+    if (refreshCallback == nullptr) {
+        valid_authentication = false;
+        syslog(LOG_ERR, "OAuthTokenManager was supplied an invalid refresh callback.");
+        printf("OAuthTokenManager was supplied an invalid refresh callback.");
+        return;
+    }
+
+    httpClient = std::make_shared<CurlEasyClient>(constants::max_concurrency_oauth);
+    refreshTokenCallback = refreshCallback;
+
+    try {
+        refresh_token();
+    } catch(std::exception& ex) {
+        syslog(LOG_ERR, "Unable to retrieve OAuth token: %s", ex.what());
+        printf("Unable to retrieve OAuth token: %s\n", ex.what());
+        valid_authentication = false;
+        return;
+    }
+
+    if(current_oauth_token.empty()) {
+        valid_authentication = false;
+        syslog(LOG_ERR, "Unable to retrieve OAuth Token with given credentials.");
+        printf("Unable to retrieve OAuth Token with given credentials.\n");
+    }
+}
+
+/// <summary>
+/// Check for valid authentication which is set by the constructor, and refresh functions.
+/// </summary>
+bool OAuthTokenCredentialManager::is_valid_connection()
+{
+    return valid_authentication;
+}
+
+/// <summary>
+/// Refreshes the token. Note this can throw an error, so be prepared to catch.
+/// Unless you absolutely _need_ to force a refresh, just call get_token instead.
+/// Furthermore, note that refresh_token does not manage the token lock itself-- Please lock the token if you plan to explicitly refresh.
+/// <param>
+/// </summary>
+OAuthToken OAuthTokenCredentialManager::refresh_token()
+{
+    try {
+        current_oauth_token = refreshTokenCallback(httpClient);
+        return current_oauth_token;
+    } catch(std::exception& ex) {
+        valid_authentication = false;
+        throw ex;
+    }
+}
+
+/// <summary>
+/// Returns current oauth_token, implicitly refreshing if the current token is invalid.
+/// Note that this can throw an error if the refresh fails, so be prepared to catch.
+/// </summary>
+OAuthToken OAuthTokenCredentialManager::get_token()
+{
+    if (is_token_expired()) {
+        if (token_mutex.try_lock()) {
+            try {
+                refresh_token();
+            } catch (std::exception &ex) {

```

Comment:
>std::exception &e [](start = 21, length = 17)

Being passed through multiple exception covers like this seems to make ex.what() turn into "std::exception()". TODO: Make the actual error get dropped, rather than that.